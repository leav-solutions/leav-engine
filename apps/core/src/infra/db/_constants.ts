// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export const ARANGODB_WRITEWRITE_CONFLICT_CODE = 1200;
export const ARANGODB_HTTP_CONFLICT_CODE = 409;
