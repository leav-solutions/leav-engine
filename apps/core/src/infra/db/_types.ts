// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
// Copyright LEAV Solutions 2017
// This file is released under LGPL V3

import {AqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';

// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export interface IDbDocument {
    _id: string;
    _key: string;
    _rev: string;
    [key: string]: any;
}

export interface IDbEdge extends IDbDocument {
    _from: string;
    _to: string;
}

export interface IExecute {
    query: string | AqlQuery;
    ctx: IQueryInfos;
    withTotalCount?: boolean;
    attempts?: number;
}

export interface IExecuteWithCount<T = IDbDocument> {
    totalCount: number;
    results: T[];
}

export const isExecuteWithCount = (res: IExecuteWithCount<unknown> | unknown[]): res is IExecuteWithCount<unknown> => {
    return typeof (res as IExecuteWithCount).results !== 'undefined';
};
