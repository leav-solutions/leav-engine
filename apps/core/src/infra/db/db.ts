// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {IConfig} from '_types/config';

interface IDeps {
    config?: IConfig;
}

const _getDbConnection = (config: IConfig): Database => {
    return new Database({
        url: config.db.url
    });
};

export default function (deps: IDeps): Database {
    const db = _getDbConnection(deps.config);

    return db.database(deps.config.db.name);
}

export const initDb = async (config: IConfig) => {
    const db = _getDbConnection(config);

    const databases = await db.listDatabases();
    const dbExists = databases.reduce((exists, d) => exists || d === config.db.name, false);

    if (!dbExists) {
        await db.createDatabase(config.db.name);
    }

    db.close();
};
