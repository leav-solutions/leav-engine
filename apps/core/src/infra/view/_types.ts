// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IDbService} from 'infra/db/dbService';
import {IDbUtils} from 'infra/db/dbUtils';
import {IList} from '_types/list';
import {IGetCoreEntitiesParams} from '_types/shared';
import {IQueryInfos} from '../../_types/queryInfos';

export const VIEWS_COLLECTION_NAME = 'core_views';

export interface IViewRepo {
    createView(view: IView, ctx: IQueryInfos): Promise<IView>;
    updateView(view: IView, ctx: IQueryInfos): Promise<IView>;
    getViews(params: IGetCoreEntitiesParams, ctx: IQueryInfos): Promise<IList<IView>>;
    deleteView(viewId: string, ctx: IQueryInfos): Promise<IView>;
}

export interface IViewRepoDeps {
    'core.infra.db.dbService': IDbService;
    'core.infra.db.dbUtils': IDbUtils;
}
