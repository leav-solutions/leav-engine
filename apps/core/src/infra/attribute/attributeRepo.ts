// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql} from 'arangojs';
import {IUtils} from 'utils/utils';
import {IList} from '_types/list';
import {IQueryInfos} from '_types/queryInfos';
import {IAttribute, IGetCoreAttributesParams} from '../../_types/attribute';
import {IGetCoreEntitiesParams} from '../../_types/shared';
import {IDbService} from '../db/dbService';
import {IDbUtils} from '../db/dbUtils';
import {LIB_ATTRIB_COLLECTION_NAME, LIB_COLLECTION_NAME} from '../library/libraryRepo';
import {IValueRepo} from '../value/valueRepo';

export interface IAttributeRepo {
    getAttributes({params, ctx}: {params?: IGetCoreAttributesParams; ctx: IQueryInfos}): Promise<IList<IAttribute>>;
    updateAttribute({attrData, ctx}: {attrData: IAttributeForRepo; ctx: IQueryInfos}): Promise<IAttribute>;
    createAttribute({attrData, ctx}: {attrData: IAttributeForRepo; ctx: IQueryInfos}): Promise<IAttribute>;
    deleteAttribute({attrData, ctx}: {attrData: IAttribute; ctx: IQueryInfos}): Promise<IAttribute>;
    getLibraryAttributes({libraryId, ctx}: {libraryId: string; ctx: IQueryInfos}): Promise<IAttribute[]>;
    getLibraryFullTextAttributes({libraryId, ctx}: {libraryId: string; ctx: IQueryInfos}): Promise<IAttribute[]>;
}

export const ATTRIB_COLLECTION_NAME = 'core_attributes';

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.db.dbUtils'?: IDbUtils;
    'core.infra.value'?: IValueRepo;
    'core.utils'?: IUtils;
}

export interface IAttributeForRepo extends IAttribute {
    multiple_values: boolean;
}

export default function({
    'core.infra.db.dbService': dbService = null,
    'core.infra.db.dbUtils': dbUtils = null,
    'core.infra.value': valueRepo = null,
    'core.utils': utils = null
}: IDeps = {}): IAttributeRepo {
    return {
        async getLibraryAttributes({libraryId, ctx}): Promise<IAttribute[]> {
            // TODO: use aql template tag, and find out why it doesn't work :)
            const query = `
                FOR v
                IN 1 OUTBOUND '${LIB_COLLECTION_NAME}/${libraryId}'
                ${LIB_ATTRIB_COLLECTION_NAME}
                RETURN v
            `;

            const res = await dbService.execute({query, ctx});

            return res.map(a => dbUtils.cleanup<IAttribute>(a));
        },
        async getLibraryFullTextAttributes({libraryId, ctx}): Promise<IAttribute[]> {
            const libAttributesCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);
            const attributesCollec = dbService.db.edgeCollection(ATTRIB_COLLECTION_NAME);

            const attrs = await dbService.execute({
                query: aql`LET fullTextAttrs = (
                            FOR e IN ${libAttributesCollec}
                                FILTER e._from == ${LIB_COLLECTION_NAME + '/' + libraryId}
                                FILTER e.full_text_search == true
                            RETURN LAST(SPLIT(e._to, '/'))
                        )
                        FOR a IN ${attributesCollec}
                            FILTER POSITION(fullTextAttrs, a._key)
                        RETURN a
                    `,
                ctx
            });

            return attrs.map(a => dbUtils.cleanup<IAttribute>(a));
        },
        async getAttributes({
            params,
            ctx
        }: {
            params?: IGetCoreAttributesParams;
            ctx: IQueryInfos;
        }): Promise<IList<IAttribute>> {
            const _generateLibrariesFilterConds = (filterKey: string, filterVal: string | boolean | string[]) => {
                if (typeof filterVal === 'boolean') {
                    return aql``;
                }

                const libs = utils.forceArray(filterVal);

                const valParts = libs.map(l => aql`v._key == ${l}`);
                const libKeyCond = aql.join(valParts, ' OR ');

                // Check if there is links between given libraries and attribute
                return aql`LENGTH(
                    FOR v IN 1 INBOUND el
                    core_edge_libraries_attributes
                    FILTER ${libKeyCond}
                    RETURN v._key
                ) > 0`;
            };

            const defaultParams: IGetCoreEntitiesParams = {
                filters: null,
                strictFilters: false,
                withCount: false,
                pagination: null,
                sort: null
            };

            const initializedParams = {...defaultParams, ...params};

            return dbUtils.findCoreEntity<IAttribute>({
                ...initializedParams,
                collectionName: ATTRIB_COLLECTION_NAME,
                customFilterConditions: {libraries: _generateLibrariesFilterConds},
                ctx
            });
        },
        async updateAttribute({attrData, ctx}): Promise<IAttribute> {
            const docToInsert = dbUtils.convertToDoc(attrData);

            // Insert in libraries collection
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);
            const res = await dbService.execute({
                query: aql`UPDATE ${docToInsert} IN ${col} RETURN NEW`,
                ctx
            });

            return dbUtils.cleanup(res.pop());
        },
        async createAttribute({attrData, ctx}): Promise<IAttribute> {
            const docToInsert = dbUtils.convertToDoc(attrData);
            // Insert in libraries collection
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);
            const res = await dbService.execute({
                query: aql`INSERT ${docToInsert} IN ${col} RETURN NEW`,
                ctx
            });

            return dbUtils.cleanup(res.pop());
        },
        async deleteAttribute({attrData, ctx}): Promise<IAttribute> {
            // Delete links library<->attribute
            const libAttributesCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            // Delete all values
            await valueRepo.clearAllValues({attribute: attrData, ctx});

            await dbService.execute({
                query: aql`
                    FOR e IN ${libAttributesCollec}
                        FILTER e._to == ${'core_attributes/' + attrData.id}
                        REMOVE e IN ${libAttributesCollec}
                    `,
                ctx
            });

            // Delete attribute
            const col = dbService.db.collection(ATTRIB_COLLECTION_NAME);

            const res = await dbService.execute({
                query: aql`REMOVE ${{_key: attrData.id}} IN ${col} RETURN OLD`,
                ctx
            });

            // Return deleted attribute
            return dbUtils.cleanup(res.pop());
        }
    };
}
