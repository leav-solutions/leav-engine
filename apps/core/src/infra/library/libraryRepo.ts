// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql} from 'arangojs';
import {IElasticsearchService} from 'infra/elasticsearch/elasticsearchService';
import {difference} from 'lodash';
import {IQueryInfos} from '_types/queryInfos';
import {IGetCoreEntitiesParams} from '_types/shared';
import {IAttributeFilterOptions} from '../../_types/attribute';
import {ILibrary} from '../../_types/library';
import {IList} from '../../_types/list';
import {IAttributeRepo} from '../attribute/attributeRepo';
import {IDbService} from '../db/dbService';
import {IDbUtils} from '../db/dbUtils';

export const LIB_COLLECTION_NAME = 'core_libraries';
export const LIB_ATTRIB_COLLECTION_NAME = 'core_edge_libraries_attributes';

export interface ILibraryRepo {
    /**
     * Return libraries
     *
     * filters                   Filters libraries returned
     *
     * @return Promise<Array<object>>   All libraries data
     */
    getLibraries({params, ctx}: {params?: IGetCoreEntitiesParams; ctx: IQueryInfos}): Promise<IList<ILibrary>>;

    /**
     * Create new library
     *
     * @return Promise<object>  New library data
     */
    createLibrary({libData, ctx}: {libData: ILibrary; ctx: IQueryInfos}): Promise<ILibrary>;

    /**
     * Update existing library
     *
     * libData   Must contain "id" key to identify library to update
     *
     * @return object   Updated library data
     */
    updateLibrary({libData, ctx}: {libData: ILibrary; ctx: IQueryInfos}): Promise<ILibrary>;

    /**
     * Delete library
     *
     * @return {}   Deleted library data
     */
    deleteLibrary({id, ctx}: {id: string; ctx: IQueryInfos}): Promise<ILibrary>;

    /**
     * Link attributes to library
     *
     * attributes Array of attributes IDs
     *
     * @return array     List of linked attributes
     */
    saveLibraryAttributes({
        libId,
        attributes,
        ctx
    }: {
        libId: string;
        attributes: string[];
        ctx: IQueryInfos;
    }): Promise<string[]>;

    /**
     * Upsert full text attributes to library
     *
     * fullTextAttributes Array of full text attributes IDs
     *
     */
    saveLibraryFullTextAttributes({
        libId,
        fullTextAttributes,
        ctx
    }: {
        libId: string;
        fullTextAttributes: string[];
        ctx: IQueryInfos;
    }): Promise<void>;

    getLibrariesUsingAttribute(attributeId: string, ctx: IQueryInfos): Promise<string[]>;
}

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.db.dbUtils'?: IDbUtils;
    'core.infra.attribute'?: IAttributeRepo;
    'core.infra.elasticsearch.elasticsearchService'?: IElasticsearchService;
}

export default function ({
    'core.infra.db.dbService': dbService = null,
    'core.infra.db.dbUtils': dbUtils = null,
    'core.infra.attribute': attributeRepo = null,
    'core.infra.elasticsearch.elasticsearchService': elasticsearchService = null
}: IDeps = {}): ILibraryRepo {
    return {
        async getLibraries({params = {}, ctx}): Promise<IList<ILibrary>> {
            const defaultParams: IGetCoreEntitiesParams = {
                filters: null,
                strictFilters: false,
                withCount: false,
                pagination: null,
                sort: null
            };

            const initializedParams = {...defaultParams, ...params};
            const libraries = await dbUtils.findCoreEntity<ILibrary>({
                ...initializedParams,
                collectionName: LIB_COLLECTION_NAME,
                ctx
            });

            return libraries;
        },
        async createLibrary({libData, ctx}): Promise<ILibrary> {
            const docToInsert = dbUtils.convertToDoc(libData);

            // Create new collection for library
            await dbService.createCollection(docToInsert._key);

            // Insert in libraries collection
            const libCollc = dbService.db.collection(LIB_COLLECTION_NAME);
            const libRes = await dbService.execute({
                query: aql`INSERT ${docToInsert} IN ${libCollc} RETURN NEW`,
                ctx
            });

            // Create elasticsearch index
            if (!(await elasticsearchService.indiceExists(libData.id))) {
                await elasticsearchService.indiceCreate(libData.id);
            }

            return dbUtils.cleanup(libRes.pop());
        },
        async updateLibrary({libData, ctx}): Promise<ILibrary> {
            const docToInsert = dbUtils.convertToDoc(libData);
            delete docToInsert.attributes; // Attributes have to be handled separately

            // Insert in libraries collection
            const col = dbService.db.collection(LIB_COLLECTION_NAME);
            const res = await dbService.execute({
                query: aql`UPDATE ${docToInsert} IN ${col} RETURN NEW`,
                ctx
            });

            return dbUtils.cleanup(res.pop());
        },
        async deleteLibrary({id, ctx}): Promise<ILibrary> {
            const filters: IAttributeFilterOptions = {linked_library: id};
            // Delete attributes linked to this library
            const linkedAttributes = await attributeRepo.getAttributes({
                params: {filters},
                ctx
            });

            for (const linkedAttribute of linkedAttributes.list) {
                await attributeRepo.deleteAttribute({
                    attrData: linkedAttribute,
                    ctx
                });
            }

            // Delete library attributes
            const libAttributesCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            await dbService.execute({
                query: aql`FOR e IN ${libAttributesCollec}
                         FILTER e._from == ${LIB_COLLECTION_NAME + '/' + id}
                         REMOVE e IN ${libAttributesCollec}`,
                ctx
            });

            // Delete library
            const col = dbService.db.collection(LIB_COLLECTION_NAME);
            const res = await dbService.execute({
                query: aql`REMOVE ${{_key: id}} IN ${col} RETURN OLD`,
                ctx
            });

            // Delete library's collection
            await dbService.dropCollection(id);

            // Return deleted library
            return dbUtils.cleanup(res.pop());
        },
        async saveLibraryAttributes({libId, attributes, ctx}): Promise<string[]> {
            // TODO: in CONCAT, query will fail is using constant instead of hard coding 'core_attributes'
            const libAttribCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            // Get current library attributes
            const currentAttrs = await attributeRepo.getLibraryAttributes({libraryId: libId, ctx});
            const deletedAttrs = difference(
                currentAttrs.filter(a => !a.system).map(a => a.id),
                attributes
            );

            // Unlink attributes not used anymore
            if (deletedAttrs.length) {
                await dbService.execute({
                    query: aql`
                        FOR attr IN ${deletedAttrs}
                            FOR l in ${libAttribCollec}
                                FILTER
                                    l._from == ${LIB_COLLECTION_NAME + '/' + libId}
                                    AND l._to == CONCAT('core_attributes/', attr)
                                REMOVE l
                                IN ${libAttribCollec}
                                RETURN OLD
                    `,
                    ctx
                });
            }

            // Save new ones
            const libAttribRes = await dbService.execute({
                query: aql`
                    FOR attr IN ${attributes}
                        LET attrToInsert = {
                            _from: ${LIB_COLLECTION_NAME + '/' + libId},
                            _to: CONCAT('core_attributes/', attr)
                        }
                        UPSERT {
                            _from: ${LIB_COLLECTION_NAME + '/' + libId},
                            _to: CONCAT('core_attributes/', attr)
                        }
                        INSERT attrToInsert
                        UPDATE attrToInsert
                        IN ${libAttribCollec}
                        RETURN NEW
                `,
                ctx
            });

            return libAttribRes.map(res => res._to.split('/')[1]);
        },
        async saveLibraryFullTextAttributes({libId, fullTextAttributes, ctx}): Promise<void> {
            const libAttribCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            await dbService.execute({
                query: aql`
                    FOR attr IN ${libAttribCollec}
                        FILTER attr._from == ${LIB_COLLECTION_NAME + '/' + libId}
                        UPDATE {
                            _key: attr._key,
                            full_text_search: POSITION(${fullTextAttributes}, LAST(SPLIT(attr._to, '/')))
                        }
                        IN ${libAttribCollec}
                `,
                ctx
            });
        },
        async getLibrariesUsingAttribute(attributeId: string, ctx: IQueryInfos): Promise<string[]> {
            const libAttributesCollec = dbService.db.edgeCollection(LIB_ATTRIB_COLLECTION_NAME);

            const res = await dbService.execute<string[]>({
                query: aql`FOR e IN ${libAttributesCollec}
                        FILTER e._to == ${'core_attributes/' + attributeId}
                    RETURN LAST(SPLIT(e._from, ${LIB_COLLECTION_NAME + '/'}))`,
                ctx
            });

            return res;
        }
    };
}
