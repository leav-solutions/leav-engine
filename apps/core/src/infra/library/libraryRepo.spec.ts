// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {AttributeTypes} from '../../_types/attribute';
import {IAttributeRepo} from '../attribute/attributeRepo';
import {IDbUtils} from '../db/dbUtils';
import libraryRepo from './libraryRepo';

describe('LibraryRepo', () => {
    const ctx = {
        userId: '0',
        queryId: '123456'
    };
    describe('getLibrary', () => {
        test('Should return all libs if no filter', async function () {
            const mockDbServ = {db: null, execute: global.__mockPromise([])};
            const mockDbUtils: Mockify<IDbUtils> = {
                findCoreEntity: global.__mockPromise([
                    {
                        id: 'users',
                        system: false,
                        label: {
                            fr: 'users'
                        }
                    }
                ])
            };

            const repo = libraryRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const trees = await repo.getLibraries({ctx});

            expect(mockDbUtils.findCoreEntity.mock.calls.length).toBe(1);
            expect(trees).toEqual([
                {
                    id: 'users',
                    system: false,
                    label: {
                        fr: 'users'
                    }
                }
            ]);
        });
    });

    describe('createLibrary', () => {
        const docLibData = {
            _key: 'test_library',
            system: true
        };
        const libData = {
            id: 'test_library',
            system: true
        };
        test('Should insert a library and create a new collection', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([docLibData]),
                createCollection: global.__mockPromise()
            };

            const mockCleanupRes = libData;
            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValue(mockCleanupRes),
                convertToDoc: jest.fn().mockReturnValue(docLibData)
            };

            const libRepo = libraryRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const createdLib = await libRepo.createLibrary({libData, ctx});
            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(mockDbServ.createCollection.mock.calls.length).toBe(1);

            // First call is to insert library
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/^INSERT/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(createdLib).toMatchObject(libData);
        });
    });

    describe('updateLibrary', () => {
        const docLibData = {_key: 'test_library', system: true};
        const libData = {id: 'test_library', system: true};
        test('Should update library', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise([docLibData])
            };

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValue(libData),
                convertToDoc: jest.fn().mockReturnValue(docLibData)
            };

            const libRepo = libraryRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils
            });

            const updatedLib = await libRepo.updateLibrary({libData, ctx});

            expect(mockDbServ.execute.mock.calls.length).toBe(1);

            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/^UPDATE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            expect(updatedLib).toMatchObject(libData);
        });
    });

    describe('deleteLibrary', () => {
        const libData = {id: 'test_lib', system: false, label: {fr: 'Test'}};

        const docLibData = {
            _key: 'test_lib',
            _id: 'core_libraries/test_lib',
            _rev: '_WSgDYea--_',
            label: {en: 'Test', fr: 'Test'},
            system: false
        };

        test('Should delete a library and return deleted library', async function () {
            const mockAttrRepo: Mockify<IAttributeRepo> = {
                getAttributes: global.__mockPromise({
                    list: [
                        {id: 'attr1', type: AttributeTypes.SIMPLE},
                        {id: 'attr2', type: AttributeTypes.SIMPLE}
                    ],
                    totalCount: 0
                }),
                deleteAttribute: global.__mockPromise({})
            };

            const mockDbServ = {
                db: new Database(),
                dropCollection: global.__mockPromise(),
                execute: global.__mockPromiseMultiple([[], [docLibData]])
            };

            const mockDbUtils: Mockify<IDbUtils> = {
                cleanup: jest.fn().mockReturnValue(libData),
                convertToDoc: jest.fn().mockReturnValue(docLibData)
            };

            const libRepo = libraryRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.db.dbUtils': mockDbUtils as IDbUtils,
                'core.infra.attribute': mockAttrRepo as IAttributeRepo
            });
            libRepo.getLibraries = global.__mockPromise([libData]);

            await libRepo.deleteLibrary({id: libData.id, ctx});

            expect(mockAttrRepo.getAttributes.mock.calls.length).toBe(1);
            expect(mockAttrRepo.deleteAttribute.mock.calls.length).toBe(2);

            expect(mockDbServ.dropCollection.mock.calls.length).toBe(1);
            expect(mockDbServ.execute.mock.calls.length).toBe(2);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/^FOR/);
            expect(mockDbServ.execute.mock.calls[1][0].query.query).toMatch(/^REMOVE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();
        });
    });

    describe('saveLibraryAttributes', () => {
        test('Should link attributes to a library and return linked attributes', async function () {
            const mockQueryRes = [
                {
                    _key: '222400216',
                    _id: 'core_edge_libraries_attributes/222400216',
                    _from: 'core_libraries/users',
                    _to: 'core_attributes/id',
                    _rev: '_WSse-um--_'
                },
                {
                    _key: '222400220',
                    _id: 'core_edge_libraries_attributes/222400220',
                    _from: 'core_libraries/users',
                    _to: 'core_attributes/my_attr',
                    _rev: '_WSse-um--B'
                }
            ];
            const mockDbServ = {
                db: new Database(),
                execute: global.__mockPromise(mockQueryRes)
            };

            const mockAttrRepo: Mockify<IAttributeRepo> = {
                getLibraryAttributes: global.__mockPromise([
                    {
                        id: 'id',
                        format: 'text',
                        label: {en: 'ID'},
                        system: true,
                        type: 'link'
                    },
                    {
                        id: 'other_attr',
                        format: 'numeric',
                        label: {en: 'Attr'},
                        system: false,
                        type: 'index'
                    },
                    {
                        id: 'created_at',
                        format: 'numeric',
                        label: {en: 'Modification date'},
                        system: true,
                        type: 'index'
                    }
                ])
            };

            const libRepo = libraryRepo({
                'core.infra.db.dbService': mockDbServ,
                'core.infra.attribute': mockAttrRepo as IAttributeRepo
            });

            const createdAttrs = await libRepo.saveLibraryAttributes({
                libId: 'users',
                attributes: ['id', 'my_attr'],
                ctx
            });
            expect(mockDbServ.execute.mock.calls.length).toBe(2);

            // First call is to delete unused attributes
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/FOR attr/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/REMOVE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

            // Second call is to save new attributes
            expect(typeof mockDbServ.execute.mock.calls[1][0]).toBe('object'); // AqlQuery
            expect(mockDbServ.execute.mock.calls[1][0].query.query).toMatch(/FOR attr/);
            expect(mockDbServ.execute.mock.calls[1][0].query.query).toMatch(/UPSERT/);
            expect(mockDbServ.execute.mock.calls[1][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[1][0].query.bindVars).toMatchSnapshot();

            expect(createdAttrs).toEqual(['id', 'my_attr']);
        });
    });

    describe('saveLibraryFullTextAttributes', () => {
        test('Should set full text attributes of a library and return full text attributes', async function () {
            const mockDbServ = {
                db: new Database(),
                execute: jest.fn()
            };

            const libRepo = libraryRepo({
                'core.infra.db.dbService': mockDbServ
            });

            await libRepo.saveLibraryFullTextAttributes({
                libId: 'users',
                fullTextAttributes: ['id'],
                ctx
            });

            expect(mockDbServ.execute.mock.calls.length).toBe(1);
            expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery

            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/FOR attr/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/UPDATE/);
            expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
            expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();
        });
    });
});
