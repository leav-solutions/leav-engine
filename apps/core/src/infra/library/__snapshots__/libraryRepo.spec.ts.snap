// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LibraryRepo createLibrary Should insert a library and create a new collection 1`] = `"INSERT @value0 IN @@value1 RETURN NEW"`;

exports[`LibraryRepo createLibrary Should insert a library and create a new collection 2`] = `
{
  "@value1": "core_libraries",
  "value0": {
    "_key": "test_library",
    "system": true,
  },
}
`;

exports[`LibraryRepo deleteLibrary Should delete a library and return deleted library 1`] = `
"FOR e IN @@value0
                         FILTER e._from == @value1
                         REMOVE e IN @@value0"
`;

exports[`LibraryRepo deleteLibrary Should delete a library and return deleted library 2`] = `
{
  "@value0": "core_edge_libraries_attributes",
  "value1": "core_libraries/test_lib",
}
`;

exports[`LibraryRepo saveLibraryAttributes Should link attributes to a library and return linked attributes 1`] = `
"
                        FOR attr IN @value0
                            FOR l in @@value1
                                FILTER
                                    l._from == @value2
                                    AND l._to == CONCAT('core_attributes/', attr)
                                REMOVE l
                                IN @@value1
                                RETURN OLD
                    "
`;

exports[`LibraryRepo saveLibraryAttributes Should link attributes to a library and return linked attributes 2`] = `
{
  "@value1": "core_edge_libraries_attributes",
  "value0": [
    "other_attr",
  ],
  "value2": "core_libraries/users",
}
`;

exports[`LibraryRepo saveLibraryAttributes Should link attributes to a library and return linked attributes 3`] = `
"
                    FOR attr IN @value0
                        LET attrToInsert = {
                            _from: @value1,
                            _to: CONCAT('core_attributes/', attr)
                        }
                        UPSERT {
                            _from: @value1,
                            _to: CONCAT('core_attributes/', attr)
                        }
                        INSERT attrToInsert
                        UPDATE attrToInsert
                        IN @@value2
                        RETURN NEW
                "
`;

exports[`LibraryRepo saveLibraryAttributes Should link attributes to a library and return linked attributes 4`] = `
{
  "@value2": "core_edge_libraries_attributes",
  "value0": [
    "id",
    "my_attr",
  ],
  "value1": "core_libraries/users",
}
`;

exports[`LibraryRepo saveLibraryFullTextAttributes Should set full text attributes of a library and return full text attributes 1`] = `
"
                    FOR attr IN @@value0
                        FILTER attr._from == @value1
                        UPDATE {
                            _key: attr._key,
                            full_text_search: POSITION(@value2, LAST(SPLIT(attr._to, '/')))
                        }
                        IN @@value0
                "
`;

exports[`LibraryRepo saveLibraryFullTextAttributes Should set full text attributes of a library and return full text attributes 2`] = `
{
  "@value0": "core_edge_libraries_attributes",
  "value1": "core_libraries/users",
  "value2": [
    "id",
  ],
}
`;

exports[`LibraryRepo updateLibrary Should update library 1`] = `"UPDATE @value0 IN @@value1 RETURN NEW"`;

exports[`LibraryRepo updateLibrary Should update library 2`] = `
{
  "@value1": "core_libraries",
  "value0": {
    "_key": "test_library",
    "system": true,
  },
}
`;
